### ISSUE TITLE
.o files > 2gb seem to fail on macOS

### ISSUE THREAD

**Issue by sandercox** [ccache/ccache#1389]
### How to reproduce ###
1. Have a compiler generate a large file of >2gb (2.1gb in our case)
2. use that with CMake and cmake stops

### Actual behavior ###
Compiling the file ends with an error:
```
ccache: error: Failed to read <path to file>.cpp.o: Invalid argument
```

### Expected behavior ###
No error from ccache.

### Environment ###

ccache version 4.9

macOS 10.12  
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
cmake 3.28.1

**Comment by jrosdahl** [comment]
Sounds like you're using a ccache built in 32-bit mode? Have you built it yourself or did you get a prebuilt version?

That said, the ccache build system sets [`_FILE_OFFSET_BITS=64`](https://github.com/ccache/ccache/blob/a2bf2956d78599b2397081a9e7b253a569c7c932/cmake/config.h.in#L58-L61) to be able to handle large files even when built in 32-bit mode. I don't know if that works on macOS or if there is another way to opt-in to 64-bit APIs.

**Comment by sandercox** [comment]
> Sounds like you're using a ccache built in 32-bit mode? Have you built it yourself or did you get a prebuilt version?
> 
> That said, the ccache build system sets [`_FILE_OFFSET_BITS=64`](https://github.com/ccache/ccache/blob/a2bf2956d78599b2397081a9e7b253a569c7c932/cmake/config.h.in#L58-L61) to be able to handle large files even when built in 32-bit mode. I don't know if that works on macOS or if there is another way to opt-in to 64-bit APIs.

Our fleet of computers uses homebrew `brew install ccache` but I'll give a manual ccache built a try to see if I can get that to properly handle 64bit files.

Thanks for info on where to my investigation.

**Comment by sandercox** [comment]
Tried a build with a manual compiled ccache
```
cmake -G Ninja -B build && cmake --build
ln -s "$(pwd)/build/ccache" /usr/local/bin/ccache
```

But that has the same invalid argument error. THis happens after about 3 minutes compiling the file in question - any pointers on where I should add a breakpoint when running a debugger for this? Happy to dig a bit to get the issue resolved!

**Comment by jrosdahl** [comment]
@sandercox: I might have fixed this in a251091f796a2b67d07ad5b0fa60d1fbc8b5cbff. Could you try if latest master works better for you?

**Comment by sandercox** [comment]
@jrosdahl Thank you for fixing this, the latest master works on that version of our code. We've also went ahead and done some work on limiting our object file size to begin with, but thank you for taking the time to fix it.
